waspc/cli/Command/Watch.hs

```diff
@@ -35,19 +44,45 @@ watch waspProjectDir outDir = FSN.withManager $ \mgr -> do
   chan <- newChan
   _ <- FSN.watchDirChan mgr (SP.toFilePath waspProjectDir) eventFilter chan
   _ <- FSN.watchTreeChan mgr (SP.toFilePath $ waspProjectDir </> Common.extCodeDirInWaspProjectDir) eventFilter chan
-  listenForEvents chan currentTime
+  timeOfLastEvent <- MVar.newMVar currentTime
+  listenForEvents chan currentTime timeOfLastEvent
   where
-    listenForEvents :: Chan FSN.Event -> UTCTime -> IO ()
-    listenForEvents chan lastCompileTime = do
+    oneSecond :: Int
+    oneSecond = 1000000
+
+    oneSecondDelay :: MVar.MVar UTCTime -> IO ()
+    oneSecondDelay timeOfLastEventMVar = do
+      threadDelay oneSecond
+      currentDelayTime <- MVar.readMVar timeOfLastEventMVar
+      currentTime <- getCurrentTime
+      let timeDiff = nominalDiffTimeToSeconds $ diffUTCTime currentTime currentDelayTime
+      when (timeDiff < fromIntegral oneSecond) $ do
+        threadDelay (oneSecond - (floor . (* 1e9) $ timeDiff))
+        oneSecondDelay timeOfLastEventMVar
+
+    recurEventCheck :: Chan FSN.Event -> UTCTime -> MVar.MVar UTCTime -> IO ()
+    recurEventCheck chan currentTime timeOfLastEventMVar = do
+      event <- readChan chan
+      let eventTime = FSN.eventTime event
+      unless (eventTime < currentTime) $ do
+        MVar.putMVar timeOfLastEventMVar eventTime
+
+    listenForEvents :: Chan FSN.Event -> UTCTime -> MVar.MVar UTCTime -> IO ()
+    listenForEvents chan lastCompileTime timeOfLastEventMVar = do
       event <- readChan chan
       let eventTime = FSN.eventTime event
       if eventTime < lastCompileTime
-        then -- If event happened before last compilation started, skip it.
-          listenForEvents chan lastCompileTime
+        then do
+          -- If event happened before last compilation started, skip it.
+          MVar.putMVar timeOfLastEventMVar lastCompileTime
+          listenForEvents chan lastCompileTime timeOfLastEventMVar
         else do
           currentTime <- getCurrentTime
+          MVar.putMVar timeOfLastEventMVar currentTime
```

Comment:
You shouldn't set this to `currentTime`, you should set it to `eventTime` -> it is the time of the last event.